RFC stage 3: `AGENTS.md` and agent steering (#143)
nf-core/proposals: proposals/0141-agent-steering.md

Concerns have been rightly raised about keeping the main steering file up to date, given the limited template release schedule and the dynamic nature of the AI sector. To address that, we propose to introduce a minimal `AGENTS.md`, consisting mostly of references to other, more updateable files.

Some community members have argued that `CONTRIBUTING.md` is a sufficient file to guide agents as well as human programmers. While this file is valuable, it is not automatically introduced into agent context, and thus should be explicitly included by reference. Furthermore, it is optimized for humans, so agent guidance should extend beyond this file.

There was previously an RFC on this topic, but it has gone stale months ago.

# Goals

- Add a minimal `AGENTS.md` to the template, with references to `CONTRIBUTING.md` and other steering files.
- Add a general `AGENTS_nfcore.md` to a public repository (website or otherwise), referenced by `AGENTS.md`.

> - Add a pipeline-specific ~~`AGENTS_[pipeline].md`~~ **section** to the template, ~~initially near-empty, intended for pipeline-specific coding standards and referenced by~~ **within** the main `AGENTS.md`. \[Alteration by IT, introduced following discussion in the original proposal issue\]

Reviewer: I think it's fine to have a cleaned up version here. Could add a sub-point mentioning that we discussed it if you want, but I'm not sure it's needed.

```suggestion
- Add a pipeline-specific section to the template, within the main `AGENTS.md`. The section will initially be nearly-empty, intended for developers to add pipeline-specific coding standards.
```

Reviewer 2: Agree with this comment --similarly to CONTRIBUTING.md pipeline-specific section after the template end.

# Non-Goals

- Overload the template with extensive agent steering (e.g. skills); this should be handled with an extension.
- Create lag and/or undue maintenance burden; all steering files should be updateable with a simple PR.

# Detailed Design

## `nf-core/agent-steering`

> To create a dedicated, findable space for agent-related documents in nf-core, a new repository called `nf-core/agent-steering` \[name open to discussion - IT\] is created. Within this RFC, a single file called `AGENTS.md` (see _infra_) will be created in the repository, alongside a standard README and LICENSE. No other file should be created, unless called for by a future RFC.

Reviewer: Add these things as comments on the PR, this is where the discussion happens.

```suggestion
To create a dedicated, findable space for agent-related documents in nf-core, a new repository called `nf-core/agent-steering` is created. Within this RFC, a single file called `AGENTS.md` (see _infra_) will be created in the root of the repository.
```

Reviewer: To be honest I'm not sure what this section adds beyond the one below, maybe just remove it.

## Template `AGENTS.md`

A new file called `AGENTS.md` is added to the nf-core pipeline template. This file contains 2 sections: the preamble and the pipeline section.

### Preamble

The preamble contains:

- a one-sentence reference to nf-core ("This repository was created with the nf-core pipeline template" or OWTTE),
- the instruction to consult `CONTRIBUTING.md`,
- the instruction to consult the remote steering file (_infra_).

The preamble does not contain any other information or instructions.

### Pipeline-specific section

The pipeline-specific section contains any standards and practices specific to a given pipeline. Developing it and keeping it up to date is the resposibility of the maintainers of the pipeline. This section should not recapitulate any general nf-core standards.

> `nf-core/modules` may utilize this section to explain the specifics of the repository.

Reviewer: I think this confuses things a bit, as `nf-core/modules` doesn't use the pipeline template.

Instead, better to have a separate section of the RFC specifically for non-pipeline nf-core repositories. Modules is one of them, see also website, tools, even proposals 😅  All could / perhap should have a comparable `AGENTS.md` but it'll have to be created one by one.

## Remote `AGENTS.md`

> A new file called `AGENTS.md` is created inside `nf-core/agent-steering` (_supra_). This file contains ample guidance for AI agents about the template and nf-core coding practices. The file will be updated as agents evolve, with updates via PR as required.

Reviewer: ```suggestion
A new file called `AGENTS.md` is created inside `nf-core/agent-steering` (_supra_). This file contains guidance for AI agents about the template and nf-core coding practices. The file will be updated as agents evolve, with updates via PR as required.
```

This RFC does not specify a rigid structure or contents, but the following elements will tentatively be included (in some order):

- a treemap of the template with explanations of directories and files,
- a description of nf-core tools, list of commands, and examples for the most common operations in pipeline development,
- a description of nf-test and basic testing concepts,
- commit and PR discipline (scopes, message format),

> - pre-commit routine (linting, tests, prek),

Reviewer 2: both nf-core and nextflow strict sntax lints

- self-disclosure rules for fully autonomous agents,

> - references to nf-core documentation as required.

Reviewer: It might be nice for the file to be fairly minimal and have separate skill files for all of these things, revealed with progressive disclosure. Some may be better as dedicated skills.

Not sure that the RFC really has to go into this detail, but thought I'd mention it here whilst thinking about it.

Author: Skills are deliberately out of scope for this RFC, but I expect them to be considered in some fashion soon

## Pipeline linting tests

A new linting test is added to `nf-core pipelines lint`. This test verifies that the preamble of `AGENTS.md` exactly matches that in the template.

## Documentation

- A paragraph about the `AGENTS.md` option is added to the documentation for `nf-core pipelines create`.
- A blog post about the changes will be published after the drafting and testing stages are concluded.
- Template `AGENTS.md` contains one-line comments for each section

# Drawbacks

- **More files in the template.** The pipeline template is heavy as it is. This RFC requires the addition of one extra top-level file, which may cause additional confusion, especially for newer contributors.

# Alternatives

- **Status quo: no centralized agent guidance.** This is the easiest solution to implement, but it is and will keep causing divergent development practices. It also increases the influx of low-quality PRs from developers who forget to include any nf-core guidance, often due to the lack of domain knowledge.

> - **Delegate all agent guidance to an extension.** While this would prevent any template pollution, it has the significant drawback of requiring an additional manual step from the developer. Thus, it is not a complete solution to the problem of poorly curated AI contributions. A more comprehensive nf-core extension might be created in the future, but is out of scope of this RFC.

Reviewer: ```suggestion
- **Delegate all agent guidance to an extension / plugin.** While this would prevent any template pollution, it has the significant drawback of requiring the developer to install the plugin ahead of time. Thus, it is not a complete solution to the problem of poorly curated AI contributions. A more comprehensive nf-core extension might be created in the future, but is out of scope of this RFC.
```

- [ ] Creating `nf-core/agent-steering` (core team)
- [ ] Pipeline `AGENTS.md` first draft (anyone)
- [ ] Central `AGENTS.md` first draft (anyone)

## Testing stage

- [ ] Manual addition of `AGENTS.md` to:
  - [ ] `nf-core/modules` (maintainers)
  - [ ] at least 5 pipelines (volunteer pipeline maintainers)
- [ ] Testing by volunteers for at least 1 month (volunteer developers)

## Implementation stage

- [ ] Pipeline `AGENTS.md` second draft (anyone)
- [ ] Central `AGENTS.md` second draft (anyone)
- [ ] Addition of `AGENTS.md` to the template (infrastructure team)
- [ ] Addition of relevant options to pipeline creation (infrastructure team)
- [ ] Addition of `AGENTS.md` lint test (infrastructure team)
- [ ] Addition of `AGENTS.md` documentation to the website (anyone)
- [ ] Blog post (anyone)
- [ ] Bytesize explaining the setup (Igor/Phil/Evangelos)

## Release stage

- [ ] tools release with the updated template (infrastructure team)
- [ ] Central `AGENTS.md` ready for swift updates (core team)

## Tentative timeline

Start of drafting: immediately
End of drafting and start of testing: by June 30, 2026
End of testing and start of implementation: by July 31, 2026

> Tools release with new features: September-October 2026

Reviewer: ```suggestion
- Start of drafting: immediately
- End of drafting and start of testing: by June 30, 2026
- End of testing and start of implementation: by July 31, 2026
- Tools release with new features: September-October 2026
```

nf-core currently has no specific `AGENTS.md` or other agent steering files. This requires extra work from developers to ensure that the agent output complies with nf-core standards, and creates room for divergence in practices. To promote uniformity and reduce the issue of non-compliant AI contributions, it would be prudent to introduce some basic guidance into the template, so that it affects every project automatically after the next update.

Concerns have been rightly raised about keeping the main steering file up to date, given the limited template release schedule and the dynamic nature of the AI sector. To address that, we propose to introduce a minimal `AGENTS.md`, consisting mostly of references to other, more updateable files.

Some community members have argued that `CONTRIBUTING.md` is a sufficient file to guide agents as well as human programmers. While this file is valuable, it is not automatically introduced into agent context, and thus should be explicitly included by reference. Furthermore, it is optimized for humans, so agent guidance should extend beyond this file.

> There was previously an RFC on this topic, but it has gone stale months ago.

Reviewer 3: Might be nice to reference it
```suggestion
There was previously an [RFC](https://github.com/nf-core/proposals/issues/61) on this topic, but it has gone stale months ago and thus "closed as not planned".
```

Reviewer 3: I saw now that is referenced at the end, but maybe still we can keep it the link here

- Add a general `AGENTS_nfcore.md` to a public repository (website or otherwise), referenced by `AGENTS.md`.
- Add a pipeline-specific section to the template, within the main `AGENTS.md`. The section will initially be nearly-empty, intended for developers to add pipeline-specific coding standards.

# Non-Goals

- Overload the template with extensive agent steering (e.g. skills); this should be handled with an extension.
- Create lag and/or undue maintenance burden; all steering files should be updateable with a simple PR.

# Detailed Design

## `nf-core/agents` and remote `AGENTS.md`

To create a dedicated, findable space for agent-related documents in nf-core, a new repository called `nf-core/agents` is created. Within this RFC, a single file called `AGENTS.md` is created in the root of the repository.

`AGENTS.md` contains guidance for AI agents about the template and nf-core coding practices. The file will be updated as agents evolve, with updates via PR as required.

This RFC does not specify a rigid structure or contents, but the following elements will tentatively be included (in some order):

> - a treemap of the template with explanations of directories and files,

Reviewer 3: I have a couple of suggestions here, but not sure whether they are already superficially covered by this point for the two first items (treemap):
* How the different configs relate to each other.
* How nf-core uses workflows, subworkflows and modules.
* Common Nextflow pitfalls to avoid (this actually could be a different file linked to it).
* Meta map convention (as it is very nf-core specific.
